### ISSUE TITLE
[FEATURE] Auto Differentiation Using Computational Graphs

### ISSUE THREAD

**Issue by xEcho1337** [brain4j-org/brain4j#18]
## Feature Description
Auto differentiation allows to automatically compute gradients, simplifying the implementation of layers.

## Current Alternatives
Right now, each layer has to implement a forward and backward method from scratch. Calculating the gradients and derivatives can become complicated in more complex architectures.

## ML Context
In almost every professional machine learning framework, such as PyTorch and Tensorflow, this feature is present.
